### PR TITLE
Opening Hours - Add AM/PM format

### PR DESCRIPTION
**Pull request by Skalii**

### CODE REVIEW

File: native/src/openingHoursParser.cpp [hunk 1]
```diff
@@ -7,12 +7,16 @@
 
 #include "Logging.h"
 
+#include "qlocale.h"
```
Comment (vshcherb) [inline native/src/openingHoursParser.cpp:10]: qlocale not part of dependency for core-legacy

File: native/src/openingHoursParser.cpp [hunk 2]
```diff
@@ -107,49 +111,49 @@ int getDayIndex(int i) {
...
-	std::vector<std::string> localDaysStr_;
-	for (int i = 0; i < 7; i++) {
-		char mbstr[100];
-		time.tm_wday = i;
-		std::strftime(mbstr, sizeof(mbstr), "%a", &time);
-
-		localDaysStr_.push_back(std::string(mbstr));
+void OpeningHoursParser::formatTime(int hours, int minutes, std::stringstream& b, bool appendAmPm) {
+	if (twelveHourFormatting) {
+        QTime time(hours, minutes);
```
Comment (vshcherb) [inline native/src/openingHoursParser.cpp:141]: qtime can't be used

File: native/src/openingHoursParser.cpp [hunk 3]
```diff
@@ -107,49 +109,93 @@ int getDayIndex(int i) {
...
 	int hour = minutes / 60;
 	int time = minutes - hour * 60;
-	formatTime(hour, time, b);
+	formatTime(hour, time, b, appendAmPM);
 }
 
-StringsHolder::StringsHolder() {
-	daysStr = std::vector<std::string>{"Su", "Mo", "Tu", "We", "Th", "Fr", "Sa"};
+void OpeningHoursParser::formatTime(int hours, int minutes, std::stringstream& b, bool appendAmPm) {
+    std::string hoursMinutes = OpeningHoursParser::getHoursMinutes(hours, minutes);
```
Comment (PaulStets) [inline native/src/openingHoursParser.cpp:138]: Formating: use tabs in core legacy

File: native/src/openingHoursParser.cpp [hunk 4]
```diff
@@ -2454,5 +2552,30 @@ void OpeningHoursParser::runTest() {
...
+	testInfo("15.01.2018 12:00", hours, "Will close at 1:00 PM");
+	testInfo("15.01.2018 13:10", hours, "Will open at 2:00 PM");
+	testInfo("15.01.2018 14:00", hours, "Open till 6:00 PM");
+	testInfo("15.01.2018 16:00", hours, "Will close at 6:00 PM");
+	testInfo("15.01.2018 18:10", hours, "Will open tomorrow at 9:00 AM");
+
+	// Don't write AM or PM twice for range
+	std::string string = "Mo-Fr 04:30-10:00, 07:30-23:00; Sa, Su, PH 13:30-23:00";
+	hours = parseOpenedHours(string);
+	testParsedAndAssembledCorrectly("Mo-Fr 4:30–10:00 AM, 7:30 AM–11:00 PM; Sa, Su, PH 1:30–11:00 PM", hours);
```
Comment (alex-osm) [inline native/src/openingHoursParser.cpp:2576]: Arabic and Chinese tests needed for AM/PM

File: native/src/openingHoursParser.cpp [hunk 5]
```diff
@@ -13,6 +13,11 @@ static const int CURRENT_DAY_TIME_LIMIT = -2;
 
 static StringsHolder stringsHolder;
 
+static bool twelveHourFormatting;
+static bool isAmpmOnLeft;
```
Comment (alex-osm) [inline native/src/openingHoursParser.cpp:17]: isAmpmOnLeft > isAmPmOnLeft

File: native/src/openingHoursParser.cpp [hunk 6]
```diff
@@ -1875,6 +1948,43 @@ void OpeningHoursParser::setAdditionalString(const std::string& key, const std::
 	stringsHolder.setAdditionalString(key, value);
 }
 
+//TODO: Delete?
```
Comment (alex-osm) [inline native/src/openingHoursParser.cpp:1951]: Comment?

File: native/src/openingHoursParser.cpp [hunk 7]
```diff
@@ -1855,13 +1923,18 @@ bool OpeningHoursParser::parseTime(const std::string& time, tm& dateTime) {
 		auto year = time.substr(6, 4);
 		auto hour = time.substr(11, 2);
 		auto min = time.substr(14, 2);
+        
+        auto a = atoi(year.c_str());
+        auto b = atoi(year.c_str()) - 1900;
```
Comment (alex-osm) [inline native/src/openingHoursParser.cpp:1928]: ?

File: native/src/openingHoursParser.cpp [hunk 8]
```diff
@@ -1970,11 +2080,26 @@ void OpeningHoursParser::testInfo(const std::string& time, const std::shared_ptr
...
 														 const std::shared_ptr<OpeningHours>& hours) {
 	auto assembledString = hours->toString();
+    
 	bool isCorrect = ohp_to_lowercase(assembledString) == ohp_to_lowercase(timeString);
 	OsmAnd::LogPrintf(OsmAnd::LogSeverityLevel::Warning, "%sok: Expected: \"%s\" got: \"%s\"",
 					  (!isCorrect ? "NOT " : ""), timeString.c_str(), assembledString.c_str());
 	if (!isCorrect) {
 		OsmAnd::LogPrintf(OsmAnd::LogSeverityLevel::Warning, "BUG!!!");
+        
+        OsmAnd::LogPrintf(OsmAnd::LogSeverityLevel::Warning, "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!");
```
Comment (alex-osm) [inline native/src/openingHoursParser.cpp:2090]: Looks like rubbish log

File: native/src/openingHoursParser.h [hunk 9]
```diff
@@ -647,11 +663,26 @@ struct OpeningHoursParser {
...
+
+	static void testAmPm();
+
    public:
 	OpeningHoursParser(const std::string& openingHours);
 	~OpeningHoursParser();
 
 	static void setAdditionalString(const std::string& key, const std::string& value);
+    static void setLocalizedDaysOfWeek(const std::vector<std::string>& value);
+    static void setLocalizedMounths(const std::vector<std::string>& value);
```
Comment (alex-osm) [inline native/src/openingHoursParser.h:680]: Mounths

File: native/src/openingHoursParser.h [hunk 10]
```diff
@@ -647,11 +663,26 @@ struct OpeningHoursParser {
...
+    static void setExternalTimeFormatterCallback(std::function<std::string (int hours, int minutes, bool appendAmPM)> callback);
```
Comment (alex-osm) [inline native/src/openingHoursParser.h:681]: Formatting

**Pull request merged**